draft-ietf-tls-esni.md

```diff
@@ -521,57 +655,62 @@ PaddedServerNameList.sni and ClientESNIInner.nonce to the backend
 
-If DNS is misconfigured so that a client receives ESNI keys for a
-server which is not prepared to receive ESNI, then the server will
-ignore the "encrypted_server_name" extension, as required by
-{{RFC8446}}; Section 4.1.2.  If the servers does not
-require SNI, it will complete the handshake with its default
-certificate. Most likely, this will cause a certificate name
-mismatch and thus handshake failure. Clients SHOULD NOT fall
-back to cleartext SNI, because that allows a network attacker
-to disclose the SNI. They MAY attempt to use another server
-from the DNS results, if one is provided.
+The retry mechanism repairs most such inconsistencies. If server and DNS keys mismatch,
+the server will respond with esni_retry_requested. If the server does not understand the
+"encrypted_server_name" extension at all, it will ignore it as required by {{RFC8446}};
+Section 4.1.2. Provided the server can present a certificate valid for the public name,
+the client can safely retry with updated settings, as described in {{client-behavior}}.
 
+If the public name does not verify or the retry fails, the client SHOULD NOT
+fall back to cleartext SNI, as this allows a network attacker to disclose the SNI.
+They MAY attempt to use another server from the DNS results, if one is provided.
 
 ## Middleboxes
 
 A more serious problem is MITM proxies which do not support this
 extension. {{RFC8446}}; Section 9.3 requires that
 such proxies remove any extensions they do not understand.
-This will have one of two results when connecting to the client-facing
-server:
-
-1. The handshake will fail if the client-facing server requires SNI.
-2. The handshake will succeed with the client-facing server's default
-   certificate.
-
-A Web client client can securely detect case (2) because it will result
-in a connection which has an invalid identity (most likely)
-but which is signed by a certificate which does not chain
-to a publicly known trust anchor. The client can detect this
-case and disable ESNI while in that network configuration.
-
-In order to enable this mechanism, client-facing servers SHOULD NOT
-require SNI, but rather respond with some default certificate.
-
-A non-conformant MITM proxy will forward the ESNI extension,
-substituting its own KeyShare value, with the result that
-the client-facing server will not be able to decrypt the SNI.
-This causes a hard failure. Detecting this case is difficult,
-but clients might opt to attempt captive portal detection
-to see if they are in the presence of a MITM proxy, and if
-so disable ESNI. Hopefully, the TLS 1.3 deployment experience
-has cleaned out most such proxies.
 
+A non-conformant MITM proxy which forwards the ESNI extension,
+substituting its own KeyShare value, will result in
+the client-facing server recognizing the key, but failing to decrypt
+the SNI. This causes a hard failure. Hopefully, the TLS 1.3 deployment
+experience has cleaned out most such proxies.
```

Comment:
As much as I like and agree with this sentiment, the last sentence can go.

Reply:
This too is an existing sentence from before this PR. :-P Happy to remove it though. Removed.